### ISSUE TITLE
How to encode request_data for MSG

### ISSUE THREAD

**Issue by uprightcarrion** [ottowayi/pycomm3#220]
I'm trying to send a message instruction which in studio is an array of SINTs.  I've haven't be successful yet, I saw this example of[ setting the system clock ](https://github.com/ottowayi/pycomm3/issues/85 ).  I've also seen how to encoder a single int or dint.  Could I get an example of how to encode and array for the request_data?

**Comment by ottowayi** [comment]
Yeah, you can turn any data type into an array by using `[]`, so `SINT[10].encode([1, 2, 3, ...])` should do it.  Some services require you to send the length of the array as well, so you may need to do `request_data=SINT.encode(10) + SINT[10].encode(...)`

I'm assuming you're using a generic message and not writing to a SINT array tag.  If you're writing to a tag, you'd want to use the `write` method and not a generic message.

**Comment by tdesroches** [comment]
Ok awesome I will try this, thanks for the example!

**Comment by tdesroches** [comment]
Ok I tried it and I'm not sure what I'm doing wrong.  Here is a working example in studio below,
![image](https://user-images.githubusercontent.com/30838946/180771048-d372faef-0082-4270-aa45-dad511403724.png)

![image](https://user-images.githubusercontent.com/30838946/180770939-65bd5352-67ba-4b4f-9854-4ce0508efb31.png)

I'm attempting to recreate it in python but I get the below response
![image](https://user-images.githubusercontent.com/30838946/180771323-fac22eb7-f7f6-464f-9c94-655b1ec272e3.png)

```
from pycomm3 import CIPDriver, Services, SINT, REAL


def read_festo_parameter():
    drive_path = '192.168.1.5'

    source = [
        8,1,0,0,0,
        0,0,0,0,0,
        0,0,0,0,0,
        0,0,0,0,0,
        0,0,0,0
        ]

    with CIPDriver(drive_path) as drive:
        param = drive.generic_message(
            service= 0x32,
            class_code= 0x96,
            instance=1, 
            attribute= 0x03,
            data_type=REAL,
            request_data=SINT[23].encode(source), 
            connected=False,
            unconnected_send=True,
            route_path=True,
            name='festo_param'
        )
        print(param)

read_festo_parameter()
```
I also tried adding the length as suggested above, no change.  As far as I can tell I've constructed it the same as in Studio.

**Comment by tdesroches** [comment]
Also the destination datatype is 
![image](https://user-images.githubusercontent.com/30838946/180772535-d0b84324-e874-45bd-ba50-1dc0ea3e1547.png)

**Comment by ottowayi** [comment]
`float32` should be the same as `REAL`.  What do you have in your comm path in the message in the PLC?

**Comment by tdesroches** [comment]
The comms tab is like this
![image](https://user-images.githubusercontent.com/30838946/180801889-3a7f501e-9485-41e2-b3ac-577bb7c0a43e.png)

To my surprise setting the connected=True as you suggested is working.  I don't have this enabled in the MSG in Studio though.  Does that mean I'm making some configuration mistake in the unconnected python code?

**Comment by ottowayi** [comment]
ohhh wait... I think I remember now there is a bug.  `unconnected_send` is only required if you have an additional route that needs to be forwarded on.  In this case, the IP address is the end of the route, so there is nowhere else to go on to.  You should be able to do it with `connected=False` and `unconnected_send=False`.  I'm currently working on basically a full rewrite of the library and in v2 `unconnected_send` is no longer an option, it will automatically add it when it's required.  It probably never should have been a user-facing option, since it requires a deeper understanding of the CIP protocol than what really is required for messaging (and that I now have compared to when I originally wrote this).

**Comment by tdesroches** [comment]
Ok I've confirmed what you said I set the connected and unconnected false and it still works
![image](https://user-images.githubusercontent.com/30838946/180807337-421e2f47-2bf6-497c-b97e-c3c14426f122.png)
Thank You! 

So if I had my CPU in a different slot so my path was something like 'xxx.xxx.xxx.xxx/2' then I would set unconnected_send=True?

**Comment by ottowayi** [comment]
Sort of, what you pass into the driver (`drive_path`) is treated a little differently in the `CIPDriver` vs the `LogixDriver`.  In Logix, it assumes you will be talking to a Logix PLC, which is (excluding Micro800s) always in a chassis.  It will default to `<ip>/backplane/0` if you only pass in an IP address, or `<ip>/backplane/<slot>` if you pass `<ip>/<slot>`.  In the CIPDriver, it makes no changes to the path.  So if you use the `CIPDriver` with a Logix device and only pass in the IP, you will be talking to the EN2T/etc card and not the PLC.  You will have to manually add the path to the PLC slot, `<ip>/<slot>` is an invalid path in the `CIPDriver` since the CIP path must be in port/link pairs, you'd need `<ip>/bp/<slot>` instead.    

When it comes to generic messaging, `route_path=True` means use the path from the driver, but you could also pass in a different path to talk to another device.  In your use case, it looks like you're talking directly to a drive or something?  You would need an unconnected send if you were routing.  For example, I have a Logix chassis with an EN2T at 10.1.1.100, and also in that chassis in slot 3 I have another EN2T connected to a private network with the drive I wanna talk to at 192.168.1.120.  The path would be `10.1.1.100/bp/3/2/192.168.1.120`.  `bp/3/2/192.168.1.120` is the route that needs to be forwarded on by the EN2T at 10.1.1.100, which is done by adding in the unconnected send.  So I could create a driver that opens a single connection to our entrypoint (10.1.1.100) and then from there message any number of different drives by setting the route path in the `generic_message` call.  Like:

```
with CIPDriver('10.1.1.100') as driver:

    drive_ips = ['192.168.1.120', '192.168.1.121', ...]

    for ip in drive_ips:
        param = driver.generic_message(
            ...,
            route_path=f'bp/3/2/{ip}',
            unconnected_send=True,
            connected=False,

```

What also might not be clear, if you use a connected message, then it will always use the drive's path and ignore `route_path` and `unconnected_send`.  I plan on reworking this method in v2 so that it isn't as confusing. 

TL;DR - yeah, if you have a route you need it (at least until I get v2 done).

**Comment by tdesroches** [comment]
Thanks for the details, I've been getting asked to do this kind of stuff more and more.  I learn a little more every time.
